src/hmem_ze.c

```diff
@@ -0,0 +1,200 @@
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include "ofi_hmem.h"
+#include "ofi.h"
+
+#ifdef HAVE_LIBZE
+
+#include <level_zero/ze_api.h>
+
+#define ZE_MAX_DEVICES 4
+
+static ze_driver_handle_t driver;
+static ze_device_handle_t devices[ZE_MAX_DEVICES];
+static ze_command_queue_handle_t cmd_queue[ZE_MAX_DEVICES];
+static int num_devices = 0;
+
+static inline int _ze_cmd_queue_create(ze_device_handle_t device,
+				       ze_command_queue_handle_t *cmd_queue)
+{
+	ze_command_queue_desc_t desc = {
+		.version	= ZE_COMMAND_QUEUE_DESC_VERSION_CURRENT,
+		.flags		= ZE_COMMAND_QUEUE_FLAG_NONE,
+		.mode		= ZE_COMMAND_QUEUE_MODE_SYNCHRONOUS,
+		.priority	= ZE_COMMAND_QUEUE_PRIORITY_NORMAL,
+		.ordinal	= 0,
+	};
+
+	return zeCommandQueueCreate(device, &desc, cmd_queue);
+}
+
+static inline int _ze_cmd_list_create(ze_device_handle_t device,
+				      ze_command_list_handle_t *cmd_list)
+{
+	ze_command_list_desc_t desc = {
+		.version	= ZE_COMMAND_LIST_DESC_VERSION_CURRENT,
+		.flags		= ZE_COMMAND_LIST_FLAG_NONE,
+	};
+	return zeCommandListCreate(device, &desc, cmd_list);
+}
+
+int ze_hmem_init(void)
+{
+	ze_result_t ze_ret;
+	uint32_t count;
+
+	ze_ret = zeInit(ZE_INIT_FLAG_NONE);
+	if (ze_ret)
+		goto err;
+
+	count = 1;
+	ze_ret = zeDriverGet(&count, &driver);
+	if (ze_ret)
+		goto err;
+
+	count = 0;
+	ze_ret = zeDeviceGet(driver, &count, NULL);
+	if (ze_ret || count > ZE_MAX_DEVICES)
+		goto err;
+
+	ze_ret = zeDeviceGet(driver, &count, devices);
+	if (ze_ret)
+		goto err;
+
+	num_devices = count;
+
+	for (num_devices = 0; num_devices < count; num_devices++) {
+		ze_ret = _ze_cmd_queue_create(devices[num_devices],
+					      &cmd_queue[num_devices]);
+		if (ze_ret)
+			goto free;
+	}
+
+	return FI_SUCCESS;
+
+free:
+	(void) ze_hmem_cleanup();
+
+err:
+	FI_WARN(&core_prov, FI_LOG_CORE,
+		"Failed to initialize ZE driver resources\n");
+
+	return -FI_EIO;
+}
+
+int ze_hmem_cleanup(void)
+{
+	int i, ret = FI_SUCCESS;
+
+	for (i = 0; i < num_devices; i++) {
+		if (cmd_queue[i] && zeCommandQueueDestroy(cmd_queue[i])) {
+			FI_WARN(&core_prov, FI_LOG_CORE,
+				"Failed to destroy ZE cmd_queue\n");
+			ret = -FI_EINVAL;
+		}
+	}
+
+	return ret;
+}
+
+int ze_hmem_copy(void *dst, const void *src, uint64_t device, size_t size)
+{
+	ze_command_list_handle_t cmd_list;
+	ze_result_t ze_ret;
+
+	ze_ret = _ze_cmd_list_create(devices[device], &cmd_list);
+	if (ze_ret)
+		goto err;
+
+	ze_ret = zeCommandListAppendMemoryCopy(cmd_list, dst, src, size, NULL);
+	if (ze_ret)
+		goto free;
+
+	ze_ret = zeCommandListClose(cmd_list);
+	if (ze_ret)
+		goto free;
+
+	ze_ret = zeCommandQueueExecuteCommandLists(cmd_queue[device], 1,
+						   &cmd_list, NULL);
+
+free:
+	if (!zeCommandListDestroy(cmd_list) && !ze_ret)
```

Comment:
Looks like the memcpy is async. Pardon my ze ignorance, but does zeCommandListDestroy() ensure the memcpy has completed?

Reply:
How dare you not be well-versed on an Intel API that was barely just released and is not in use? :P 
In L0, command queues can be created as either synchronous or asynchronous (see zeCommandQueueCreate). These are created synchronously which ensures the copy to be done after the call to execute. If created asynchronously, they would need a zeCommandQueueSynchronize() call.

Reply:
👍